src/endpoints/authentication/cose_auth.cpp

```diff
@@ -275,4 +372,91 @@ namespace ccf
          "Request payload must be a COSE Sign1 document, with expected "
          "protected headers. "
          "Signer must be a member identity registered with this service."}});
+
+  UserCOSESign1AuthnPolicy::UserCOSESign1AuthnPolicy() = default;
+  UserCOSESign1AuthnPolicy::~UserCOSESign1AuthnPolicy() = default;
+
+  std::unique_ptr<AuthnIdentity> UserCOSESign1AuthnPolicy::authenticate(
+    kv::ReadOnlyTx& tx,
+    const std::shared_ptr<ccf::RpcContext>& ctx,
+    std::string& error_reason)
+  {
+    const auto& headers = ctx->get_request_headers();
+    const auto content_type_it = headers.find(http::headers::CONTENT_TYPE);
+    if (content_type_it == headers.end())
+    {
+      error_reason =
+        fmt::format("Missing {} header", http::headers::CONTENT_TYPE);
+      return nullptr;
+    }
+    if (content_type_it->second != http::headervalues::contenttype::COSE)
+    {
+      error_reason = fmt::format(
+        "Content type is not set to {}", http::headervalues::contenttype::COSE);
+      return nullptr;
+    }
+
+    auto [phdr, cose_signature] =
+      cose::extract_protected_header_and_signature(ctx->get_request_body());
+
+    if (!cose::is_ecdsa_alg(phdr.alg))
+    {
+      error_reason = fmt::format("Unsupported algorithm: {}", phdr.alg);
+      return nullptr;
+    }
+
+    UserCerts users_certs_table(Tables::USER_CERTS);
+    auto user_certs = tx.ro(users_certs_table);
+    auto user_cert = user_certs->get(phdr.kid);
+    if (user_cert.has_value())
+    {
+      auto verifier = crypto::make_cose_verifier(user_cert->raw());
+
+      std::span<const uint8_t> body = {
+        ctx->get_request_body().data(), ctx->get_request_body().size()};
+      std::span<uint8_t> authned_content;
+      if (!verifier->verify(body, authned_content))
+      {
+        error_reason = fmt::format("Failed to validate COSE Sign1");
+        return nullptr;
+      }
+
+      auto identity = std::make_unique<UserCOSESign1AuthnIdentity>();
+      identity->user_id = phdr.kid;
+      identity->user_cert = user_cert.value();
+      identity->protected_header = phdr;
+      identity->envelope = body;
+      identity->content = authned_content;
+      identity->signature = cose_signature;
+      return identity;
+    }
+    else
+    {
+      error_reason = fmt::format("Signer is not a known member");
```

Comment:
```suggestion
      error_reason = fmt::format("Signer is not a known user");
```

Also report the request `kid` in the error message too maybe?

Reply:
It's provided by the caller, so I don't see the upside. If the caller is being adversarial and sets a large kid, we'll have to send a large response.